controllers/keystoneendpoint_controller.go

```diff
@@ -0,0 +1,422 @@
+
+	// update status to save current conditions to object before sub-reconcilation rules start
+	if err := r.Status().Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Handle endpoint delete
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper, os)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, instance, helper, os)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KeystoneEndpointReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&keystonev1.KeystoneEndpoint{}).
+		Complete(r)
+}
+
+func (r *KeystoneEndpointReconciler) reconcileDelete(
+	ctx context.Context,
+	instance *keystonev1.KeystoneEndpoint,
+	helper *helper.Helper,
+	os *openstack.OpenStack,
+) (ctrl.Result, error) {
+	util.LogForObject(helper, "Reconciling Endpoint delete", instance)
+
+	// only cleanup the service if there are EndpointIDs references in the
+	// object status
+	if len(instance.Status.EndpointIDs) > 0 {
+		// Delete Endpoints
+		for endpointType := range instance.Spec.Endpoints {
+			// get the gopher availability mapping for the endpointInterface
+			availability, err := openstack.GetAvailability(endpointType)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			err = os.DeleteEndpoint(
+				r.Log,
+				openstack.Endpoint{
+					Name:         instance.Spec.ServiceName,
+					ServiceID:    instance.Status.ServiceID,
+					Availability: availability,
+				},
+			)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		util.LogForObject(helper, fmt.Sprintf("Not deleting %s endpoints as there are no stored IDs", instance.Spec.ServiceName), instance)
+	}
+
+	// Service is deleted so remove the finalizer.
+	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
+	util.LogForObject(helper, "Reconciled Endpoint delete successfully", instance)
+
+	if err := r.Update(ctx, instance); err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *KeystoneEndpointReconciler) reconcileNormal(
+	ctx context.Context,
+	instance *keystonev1.KeystoneEndpoint,
+	helper *helper.Helper,
+	os *openstack.OpenStack,
+) (ctrl.Result, error) {
+	util.LogForObject(helper, "Reconciling Endpoint normal", instance)
+
+	// If the service object doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(instance, helper.GetFinalizer())
+	// Register the finalizer immediately to avoid orphaning resources on delete
+	if err := r.Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	//
+	// Wait for KeystoneService is Ready and get the ServiceID from the object
+	//
+	ksSvc, err := keystonev1.GetKeystoneServiceWithName(ctx, helper, instance.Spec.ServiceName, instance.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
I think this could hit `NotFound` errors here.  If so, would it be better to check for those and return a requeue result in such a situation?

Reply:
yes could return a specific requeue result to wait for some time and not an error. right now it returns an error and reconciles. will do

Reply:
added a commit with a change for that